jax/lax/lax.py

```diff
@@ -3952,7 +3952,131 @@ def _scatter_dtype_rule(operand, scatter_indices, updates, **kwargs):
+      raise TypeError(f"{name} in scatter op must not repeat; got: {dims}.")
+
+  update_window_dims = dimension_numbers.update_window_dims
+  inserted_window_dims = dimension_numbers.inserted_window_dims
+  scatter_dims_to_operand_dims = dimension_numbers.scatter_dims_to_operand_dims
+  # Note: in JAX, index_vector_dim is always computed as below, cf. the
+  # documentation of the ScatterDimensionNumbers class.
+  index_vector_dim = rank(scatter_indices) - 1
+
+  # This case should never happen in JAX, due to the implicit construction of
+  # index_vector_dim, but is included for completeness.
+  if rank(scatter_indices) < index_vector_dim or index_vector_dim < 0:
+    raise TypeError(f"Scatter index leaf dimension must be within [0, "
+                    f"rank(scatter_indices) + 1). rank(scatter_indices) is "
+                    f"{rank(scatter_indices)} and scatter index leaf "
+                    f"dimension is {index_vector_dim}.")
+
+  expanded_scatter_indices_shape = list(scatter_indices.shape)
+  # This case should never happen in JAX, due to the implicit construction of
+  # index_vector_dim, but is included for completeness.
+  if len(expanded_scatter_indices_shape) == index_vector_dim:
+    expanded_scatter_indices_shape.append(1)
+
+  expected_updates_rank = (len(expanded_scatter_indices_shape) - 1 +
+                           len(update_window_dims))
+
+  if rank(updates) != expected_updates_rank:
+    raise TypeError(f"Updates tensor must be of rank {expected_updates_rank}; "
+                    f"got {rank(updates)}.")
+
+  # Validate update_window_dims
+  _is_sorted(update_window_dims, "update_window_dims")
+  _no_duplicate_dims(update_window_dims, "update_window_dims")
+  _sorted_dims_in_range(update_window_dims, rank(updates), "update_window_dims")
+
+  # Validate inserted_window_dims
+  _is_sorted(inserted_window_dims, "inserted_window_dims")
+  _no_duplicate_dims(inserted_window_dims, "inserted_window_dims")
+  _sorted_dims_in_range(inserted_window_dims, rank(operand),
+                        "inserted_window_dims")
+
+  # Validate window_size
+  window_size = len(update_window_dims) + len(inserted_window_dims)
+  if rank(operand) != window_size:
+    raise TypeError(f"Scatter op has window of size {window_size}; doesn't "
+                    f"match operand of rank {rank(operand)}.")
+
+  # Validate scatter_dims_to_operand_dims
+  if (len(scatter_dims_to_operand_dims) !=
+      scatter_indices.shape[index_vector_dim]):
+    raise TypeError(f"Scatter op has {len(scatter_dims_to_operand_dims)} "
+                    f"elements in scatter_dims_to_operand_dims and the bound "
+                    f"of dimension index_vector_dim={index_vector_dim} of "
+                    f"scatter_indices is "
+                    f"{scatter_indices.shape[index_vector_dim]}. These two "
+                    f"numbers must be equal")
+
+  for i in range(len(scatter_dims_to_operand_dims)):
+    dim = scatter_dims_to_operand_dims[i]
```

Comment:
This seems to be the same check as _sorted_dims_in_range, except I assume that `scatter_dims_to_operand_dims` does not have to be sorted?

Reply:
Correct! Unless `update_window_dims` and `inserted_window_dims`, there is no assumption of being sorted for `scatter_dims_to_operand_dims` (cf the [semantics of scatter](https://www.tensorflow.org/xla/operation_semantics#scatter)).